python/src/trezorlib/tools.py

```diff
@@ -199,13 +215,15 @@ class expect:
     # Decorator checks if the method
```

Comment:
how about this:
```
M = TypeVar("M", bound=MessageType)
F = TypeVar("F")

Func = Callable[..., M]
ExpectedFunc = Callable[..., F]

@overload
def expect(expected: Type[M]) -> Callable[[Func[M]], Func[M]]:
    ...

@overload
def expect(expected: Type[M], *, type: Type[F], field: str) -> Callable[[Func[M]], Func[F]]:
    ...

def expect(expected: Type[M], *, type: Optional[Type[F]] = None, field: Optional[str] = None) -> Callable[[Func[M]], Callable]:
    def decorator(f: Func[M]) -> Callable:
        @functools.wraps(f)
        def wrapped_f(*args: Any, **kwargs: Any) -> Any:
            __tracebackhide__ = True
            ret = f(*args, **kwargs)
            if not isinstance(ret, expected):
                raise RuntimeError(...)
            if field is not None:
                return getattr(ret, field)
            else:
                return ret
        
        return wrapped_f

    return decorator
```

Reply:
then you would declare every `@expect`ed function with the true return type (so `EthereumAddress`), and stick the true return type as an argument to `@expect(field="address", type=str)`

and it _just might_ typecheck.

or maybe not.

Reply:
Thanks for suggesting this, after the latest commit it seems to work fine (however, as you mentioned, it does not check the arguments to these decorated functions)

The only mypy issue we have currently is connected with the `@session` decorator:
`src/trezorlib/tools.py:276: error: Incompatible return value type (got "Callable[[TrezorClient, VarArg(Any), KwArg(Any)], MessageType]", expected "FF")  [return-value]`